vmm/src/console_devices.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// Copyright © 2019 Intel Corporation
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+//
+
+use crate::vm_config::ConsoleOutputMode;
+use crate::Vmm;
+use libc::cfmakeraw;
+use libc::isatty;
+use libc::tcgetattr;
+use libc::tcsetattr;
+use libc::termios;
+use libc::TCSANOW;
+use std::fs::read_link;
+use std::fs::File;
+use std::fs::OpenOptions;
+use std::io;
+use std::mem::zeroed;
+use std::os::fd::AsRawFd;
+use std::os::fd::FromRawFd;
+use std::os::fd::IntoRawFd;
+use std::os::fd::RawFd;
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::net::UnixListener;
+use std::path::PathBuf;
+use std::result;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+#[cfg(target_arch = "x86_64")]
+use std::io::stdout;
+
+const TIOCSPTLCK: libc::c_int = 0x4004_5431;
+const TIOCGTPEER: libc::c_int = 0x5441;
+
+/// Errors associated with console devices
+#[derive(Debug)]
+pub enum ConsoleDeviceError {
```

Comment:
Can you evaluate these pub types to see if they can be `pub(crate)` i.e. are they only used from inside the vmm crate.

Reply:
I dropped `pub` or used `pub(crate)` to limit visibility of the elements in this module. Dropping pub in more elements causes trickling changes in other modules:

```
$ git diff
diff --git a/vmm/src/console_devices.rs b/vmm/src/console_devices.rs
index 76ae035d..8f420da4 100644
--- a/vmm/src/console_devices.rs
+++ b/vmm/src/console_devices.rs
@@ -60,7 +60,7 @@ pub enum ConsoleDeviceError {
+++ b/vmm/src/console_devices.rs
@@ -60,7 +60,7 @@ pub enum ConsoleDeviceError {
 type ConsoleDeviceResult<T> = result::Result<T, ConsoleDeviceError>;

 #[derive(Default, Clone)]
-pub struct ConsoleInfo {
+pub(crate) struct ConsoleInfo {
     // For each of File, Pty, Tty and Socket modes, below fields  hold the FD
     // of console, serial and debug devices.
     pub console_main_fd: Option<RawFd>,
diff --git a/vmm/src/device_manager.rs b/vmm/src/device_manager.rs
index d6fc1b94..34404a59 100644
--- a/vmm/src/device_manager.rs
+++ b/vmm/src/device_manager.rs
@@ -1196,7 +1196,7 @@ impl DeviceManager {
         self.console_resize_pipe.clone()
     }

-    pub fn create_devices(
+    pub(crate) fn create_devices(
         &mut self,
         console_info: Option<ConsoleInfo>,
         console_resize_pipe: Option<File>,
diff --git a/vmm/src/vm.rs b/vmm/src/vm.rs
index 71793f7b..3a290777 100644
--- a/vmm/src/vm.rs
+++ b/vmm/src/vm.rs
@@ -477,7 +477,7 @@ impl Vm {
     pub const HANDLED_SIGNALS: [i32; 1] = [SIGWINCH];

     #[allow(clippy::too_many_arguments)]
-    pub fn new_from_memory_manager(
+    pub(crate) fn new_from_memory_manager(
         config: Arc<Mutex<VmConfig>>,
         memory_manager: Arc<Mutex<MemoryManager>>,
         vm: Arc<dyn hypervisor::Vm>,
@@ -788,7 +788,7 @@ impl Vm {
     }

     #[allow(clippy::too_many_arguments)]
-    pub fn new(
+    pub(crate) fn new(
         vm_config: Arc<Mutex<VmConfig>>,
```

Please let me know if you'd like for me to pull in this change as well.

Reply:
Just do what you can - there is no need to go beyond the code you changed here.